ArduPlane/soaring.cpp

```diff
@@ -47,175 +47,134 @@ void Plane::update_soaring() {
-            gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Outside SOAR_MAX_RADIUS, RTL");
-            set_mode(mode_rtl, ModeReason::SOARING_DRIFT_EXCEEDED);
-        }
+                //update the wp location
+                g2.soaring_controller.get_target(next_WP_loc);
 
-        const char* strTooHigh = "Soaring: Too high, restoring ";
-        const char* strTooLow =  "Soaring: Too low, restoring ";
-        const char* strTooWeak = "Soaring: Thermal ended, restoring ";
-        const char* strTooFar  = "Soaring: Drifted too far, restoring ";
-
-        // Exit as soon as thermal state estimate deteriorates and we're lined up to next target
-        switch (previous_mode->mode_number()){
-        case Mode::Number::FLY_BY_WIRE_B: {
-            const bool homeIsSet = AP::ahrs().home_is_set();
-            const bool headingLinedupToHome = homeIsSet && plane.mode_loiter.isHeadingLinedUp(next_WP_loc, AP::ahrs().get_home());
-            if (homeIsSet && !headingLinedupToHome && loiterStatus!=SoaringController::LoiterStatus::ALT_TOO_LOW && timer<20e3) {
                 break;
             }
-            switch (loiterStatus) {
-                case SoaringController::LoiterStatus::ALT_TOO_HIGH:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooHigh, mode_fbwb.name());
-                    set_mode(mode_fbwb, ModeReason::SOARING_ALT_TOO_HIGH);
-                    break;
-                case SoaringController::LoiterStatus::ALT_TOO_LOW:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooLow, mode_rtl.name());
-                    set_mode(mode_rtl, ModeReason::SOARING_ALT_TOO_LOW);
-                    break;
-                default:
-                case SoaringController::LoiterStatus::THERMAL_WEAK:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooWeak, mode_fbwb.name());
-                    set_mode(mode_fbwb, ModeReason::SOARING_THERMAL_ESTIMATE_DETERIORATED);
-                    break;
-                case SoaringController::LoiterStatus::DRIFT_EXCEEDED:
-                    gcs().send_text(MAV_SEVERITY_INFO, "%s%s", strTooFar, mode_fbwb.name());
-                    set_mode(mode_fbwb, ModeReason::SOARING_DRIFT_EXCEEDED);
-                    break;
-                } // switch louterStatus
-            break;
-            }
 
-        case Mode::Number::CRUISE: {
-            const bool headingLinedupToCruise = plane.mode_loiter.isHeadingLinedUp_cd(cruise_state.locked_heading_cd);
-            if (cruise_state.locked_heading && !headingLinedupToCruise && loiterStatus!=SoaringController::LoiterStatus::ALT_TOO_LOW && timer<20e3) {
+            // Some other loiter status, we need to think about exiting loiter.
+            uint32_t timer = AP_HAL::millis() - plane.soaring_mode_timer;
+
+            if (!soaring_exit_heading_aligned() && loiterStatus!=SoaringController::LoiterStatus::ALT_TOO_LOW && timer<20e3) {
```

Comment:
We don't *usually* use exponents for this...

Also, we like to append `_ms` to times in milliseconds...

Reply:
Fixed.